### PR TITLE
[WIP] The Prototype for Multi Tenancy

### PR DESCRIPTION
**Pull request by hjianbo**
### Broken APIs/Interfaces for Users
...

### Changed Data Structures
...

### CODE REVIEW

File: apps/emqx/src/emqx_channel.erl [hunk 1]
```diff
@@ -259,6 +259,7 @@ init(
         Peercert,
         #{
             zone => Zone,
+            tenant => undefined,
```
Comment (zmstone) [inline apps/emqx/src/emqx_channel.erl:262]: -define(NO_TENANT, <<>>).

```suggestion
            tenant => ?NO_TENANT,
```

File: apps/emqx/src/emqx_channel.erl [hunk 2]
```diff
@@ -1575,6 +1577,13 @@ enrich_client(ConnPkt, Channel = #channel{clientinfo = ClientInfo}) ->
             {error, ReasonCode, Channel#channel{clientinfo = NClientInfo}}
     end.
 
+set_tenant(_ConnPkt, ConnInfo, ClientInfo = #{clientid := ClientId}) ->
+    Tenant = maps:get(peersni, ConnInfo, undefined),
```
Comment (zmstone) [inline apps/emqx/src/emqx_channel.erl:1581]:
```suggestion
    Tenant = maps:get(peersni, ConnInfo, ?NO_TENANT),
```

File: apps/emqx/src/emqx_types.erl [hunk 3]
```diff
@@ -163,7 +164,7 @@
     dn => binary(),
     atom() => term()
 }.
--type clientid() :: binary() | atom().
+-type tenant() :: maybe(binary()).
```
Comment (zmstone) [inline apps/emqx/src/emqx_types.erl:167]: use `<<>>` fo no tenant.
`undefined` is 9 bytes when serialised to external format.
```suggestion
-type tenant() :: binary().
```

File: apps/emqx/src/emqx_clientid.erl [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+-define(IS_NORMAL_ID(I), (is_atom(I) orelse is_binary(I))).
+
+-define(IS_TENANT(I), is_binary(I)).
+
+%%--------------------------------------------------------------------
+%% APIs
+%%--------------------------------------------------------------------
+
+-spec comp(emqx_types:tenant(), clientid()) -> grouped_clientid().
+comp(Tenant, ClientId) when ?IS_TENANT(Tenant), ?IS_NORMAL_ID(ClientId) ->
```
Comment (zmstone) [inline apps/emqx/src/emqx_clientid.erl:38]: maybe rename to `with_tenant`

File: apps/emqx/src/emqx_clientid.erl [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+uncomp(_) ->
+    error(badarg).
+
+-spec parse(grouped_clientid()) -> {emqx_types:tenant(), clientid()}.
+parse({Tenant, ClientId}) ->
+    {Tenant, ClientId};
+parse(ClientId) when ?IS_NORMAL_ID(ClientId) ->
+    {undefined, ClientId};
+parse(_) ->
+    error(badarg).
```
Comment (zmstone) [inline apps/emqx/src/emqx_clientid.erl:59]: one API is enough: `without_tenant` ?

File: apps/emqx/src/emqx_clientid.erl [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+update_tenant(undefined, ClientId) when ?IS_NORMAL_ID(ClientId) ->
+    ClientId;
+update_tenant(undefined, {_, ClientId}) ->
+    ClientId;
+update_tenant(Tenant, _GroupedClientId = {_, ClientId}) ->
+    {Tenant, ClientId};
+update_tenant(Tenant, ClientId) when ?IS_NORMAL_ID(ClientId) ->
+    {Tenant, ClientId};
+update_tenant(_, _) ->
+    error(badrag).
```
Comment (zmstone) [inline apps/emqx/src/emqx_clientid.erl:71]: same API? `with_tenant`

File: apps/emqx_authn/src/emqx_authn_api.erl [hunk 7]
```diff
@@ -1267,6 +1303,9 @@ paginated_list_type(Type) ->
         {meta, ref(emqx_dashboard_swagger, meta)}
     ].
 
+tenant(_Req = #{headers := Headers}) ->
+    maps:get(<<"emqx_tenant">>, Headers, undefined).
```
Comment (zmstone) [inline apps/emqx_authn/src/emqx_authn_api.erl:1307]: please make sure the headers are normalised by cowboy
```suggestion
tenant(_Req = #{headers := Headers}) ->
    maps:get(<<"emqx-tenant-id">>, Headers, undefined).
```

Comment (zmstone) [reply apps/emqx_authn/src/emqx_authn_api.erl:1307]: I still think a safer alternative is the add the tenant ID to the token to simplify the auth.
Even if for now it seems that all API calls will be issued by a deployment agent which makes sure the header is added for the intended tenant, but it feels like a security whole.

File: apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl [hunk 8]
```diff
@@ -68,7 +68,7 @@
 -type user_id() :: binary().
 
 -record(user_info, {
-    user_id :: {user_group(), user_id()},
+    user_id :: {user_group(), emqx_types:tenant(), user_id()},
```
Comment (zmstone) [inline apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl:71]: why not tenant being the first element?

File: apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl [hunk 9]
```diff
@@ -166,8 +169,9 @@ authenticate(
         password_hash_algorithm := Algorithm
     }
 ) ->
+    Tenant = maps:get(tenant, Credential, undefined),
     UserID = get_user_identity(Credential, Type),
-    case mnesia:dirty_read(?TAB, {UserGroup, UserID}) of
+    case mnesia:dirty_read(?TAB, {UserGroup, Tenant, UserID}) of
```
Comment (zmstone) [inline apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl:174]: ensure ordered_set table type.

File: apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl [hunk 10]
```diff
@@ -166,8 +169,9 @@ authenticate(
         password_hash_algorithm := Algorithm
     }
 ) ->
+    Tenant = maps:get(tenant, Credential, undefined),
     UserID = get_user_identity(Credential, Type),
-    case mnesia:dirty_read(?TAB, {UserGroup, UserID}) of
```
Comment (zmstone) [inline apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl:170]: What if we make use of the `UserGroup` for tenant?
and rename `is_superuser` to `properties` map() field, `is_superuser` being one of the properties, `group` being another.

Comment (zmstone) [reply apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl:170]: then we do not need to change the table schema, on step close to upgrade from a regular 5.0 database.

File: apps/emqx/src/emqx_ws_connection.erl [hunk 11]
```diff
@@ -376,20 +377,21 @@ get_peer_info(Type, Listener, Req, Opts) ->
         emqx_config:get_listener_conf(Type, Listener, [proxy_protocol]) andalso
             maps:get(proxy_header, Req)
     of
-        #{src_address := SrcAddr, src_port := SrcPort, ssl := SSL} ->
+        #{src_address := SrcAddr, src_port := SrcPort, ssl := SSL} = ProxyInfo ->
             SourceName = {SrcAddr, SrcPort},
             %% Notice: Only CN is available in Proxy Protocol V2 additional info
```
Comment (zmstone) [inline apps/emqx/src/emqx_ws_connection.erl:382]: this comment is apparently wrong? it also returns authority.

File: apps/emqx/src/emqx_ws_connection.erl [hunk 12]
```diff
@@ -376,20 +377,21 @@ get_peer_info(Type, Listener, Req, Opts) ->
...
             SourceSSL =
                 case maps:get(cn, SSL, undefined) of
-                    undeined -> nossl;
+                    undefined -> nossl;
```
Comment (zmstone) [inline apps/emqx/src/emqx_ws_connection.erl:385]: this fix should be ported back to `master` and `main-v4.3`

File: apps/emqx/src/emqx_channel.erl [hunk 13]
```diff
@@ -1575,6 +1577,13 @@ enrich_client(ConnPkt, Channel = #channel{clientinfo = ClientInfo}) ->
             {error, ReasonCode, Channel#channel{clientinfo = NClientInfo}}
     end.
 
+set_tenant(_ConnPkt, ConnInfo, ClientInfo = #{clientid := ClientId}) ->
```
Comment (zmstone) [inline apps/emqx/src/emqx_channel.erl:1580]: our ambition is to merge this back to master, so we'll need a global configuration to decide
either the SNI info should be used as tenant ID or not.

Comment (zmstone) [reply apps/emqx/src/emqx_channel.erl:1580]: e.g. `tenant_id_from = none | sni`

File: apps/emqx/src/emqx_ws_connection.erl (hunk 12 above)
Comment (zmstone) [reply apps/emqx/src/emqx_ws_connection.erl:385]: I have created below two tickets to follow up.
https://emqx.atlassian.net/browse/EMQX-8030
and
https://emqx.atlassian.net/browse/EMQX-8031

File: apps/emqx/src/emqx_channel.erl (hunk 13 above)
Comment (hjianbo) [reply apps/emqx/src/emqx_channel.erl:1580]: That's exactly what I want too. I will add this configuration item in subsequent commits.

File: apps/emqx/src/emqx_ws_connection.erl (hunk 11 above)
Comment (hjianbo) [reply apps/emqx/src/emqx_ws_connection.erl:382]: Does it mean that you can't get the DN through Proxy-Protocol?

**Pull request closed without merging**